core/src/main/java/org/infinispan/interceptors/impl/EntryWrappingInterceptor.java

```diff
@@ -651,40 +653,57 @@ private BasicInvocationStage setSkipRemoteGetsAndInvokeNextForDataCommand(Invoca
 
          if (trace)
             log.tracef("The return value is %s", rv);
-         if (rCtx.isInTxScope()) {
-            setSkipLookup(rCtx, dataWriteCommand.getKey());
+         if (useRepeatableRead) {
+            CacheEntry cacheEntry = rCtx.lookupEntry(dataWriteCommand.getKey());
+            // The entry is not in context when the command's execution type does not contain origin
+            if (cacheEntry != null) {
+               cacheEntry.setSkipLookup(true);
+               if (writeSkewCheck && dataWriteCommand.loadType() != VisitableCommand.LoadType.DONT_LOAD) {
+                  addVersionRead((TxInvocationContext) rCtx, cacheEntry, dataWriteCommand.getKey());
+               }
+            }
          }
       });
    }
 
    // This visitor replays the entry wrapping during remote prepare.
    // Remote writes never request the previous value from a different node,
    // so it should be safe to keep this synchronous.
+   //
+   // We ignore the ownership and wrap the entry. We have already checked writeCH for ownership
+   // - we could also check readCH, and always wrap NullCacheEntry if we cannot read but since
+   // the loaded entry is only used for listeners, it wouldn't do much good (we cannot reliably
+   // fire listeners without remote lookup).
+   // TODO: loaded entry is also used for WSC on primary and for delta/functional writes. If the node
+   // has not got the data yet, it has to request them explicitly!
```

Comment:
WSC will work as long as primary = read owner. If backup != read owner, atomic hashmaps don't work correctly, because the needed value is not fetched from remote node. In order to fix that, we would have to execute the Visitor non-synchronously and let it do a remoteGet in TxDistributionInterceptor.